Enable Custom Fields in Naming Series # 3019 of ERPNext Issue

sbkolate opened the pull request:
Hello Team,

We have developed feature to add custom field in naming series.

e.g **QTN-.YYYY.-.customer.-.####** 
Resulting in  **QTN-2016-SBK-0001** (Asuming SBK is selected customer in Quotation)

ERPNext Issue Reference: https://github.com/frappe/erpnext/issues/3019

Thanks,
Sambhaji
New Indictrans Technologies Pvt Ltd

sbkolate committed:
Enable Custom Fields in Naming Series # 3019 of ERPNext Issue

anandpdoshi commented:
Thanks for the pull request!

anandpdoshi commented on frappe/model/naming.py:
```diff
@@ -61,9 +61,8 @@ def set_name_by_naming_series(doc):
-
-def make_autoname(key, doctype=''):
+	doc.name = make_autoname(doc.naming_series+'.#####', '', doc) 
+def make_autoname(key='', doctype='', doc=''):
```
Add a blank line between 2 functions

anandpdoshi commented on frappe/model/naming.py:
```diff
@@ -109,6 +108,8 @@ def make_autoname(key, doctype=''):
 			en = today.strftime("%d")
 		elif e=='YYYY':
 			en = today.strftime('%Y')
+		elif doc and doc.get(e,None):
```
Don't need to pass 2nd arg as None. It could just be `doc and doc.get(e)`: en - doc.get(e)

anandpdoshi commented:
Should we make it as `QTN-.YYYY.-.{customer}.-.#### `, so that someone who wants to use `customer` as a literal and not a fieldname can use it. Then we check that `e.startswith("{") and e.endswith("}") and doc.get(e[1:-1]): en = doc.get(e[1:-1])`

sbkolate committed:
updated make_autoname() for adding Custom Fields in Naming Series

sbkolate commented:
Hello Anand,

I have added extra line between two function.
I have also removed _None_ from _elif doc and doc.get(e,None):_

If anyone want to use customer as literal then he can write` QTN-.YYYY.-customer-.####` i.e.need to remove dot `(.)` as we are using it to split naming_series

I am also planing to send pull request for this issue https://github.com/frappe/erpnext/issues/3813
Need to add Add example for Year Month based naming.
But since naming series is in ERPNext, I will send another pull request.

Thanks,
Sambhaji
 New Indictrans Technologies Pvt Ltd

anandpdoshi commented:
Oh yeah you are right! 

Sent from my phone

> On 19-Feb-2016, at 1:18 PM, sbkolate <notifications@github.com> wrote:
> 
> Hello Anand,
> 
> I have added extra line between two function.
> I have also removed None from elif doc and doc.get(e,None):
> 
> If anyone want to use customer as literal then he can writeQTN-.YYYY.-customer-.#### i.e.need to remove dot (.) as we are using it to split naming_series
> 
> I am also planing to send pull request for this issue frappe/erpnext#3813
> Need to add Add example for Year Month based naming.
> But since naming series is in ERPNext, I will send another pull request.
> 
> Thanks,
> Sambhaji
> New Indictrans Technologies Pvt Ltd
> 
> —
> Reply to this email directly or view it on GitHub.
>

raveslave commented:
Great to see this one. How can I see if this is deployed in release/master?

Also, did we end up using {field_name} or something else?

Tried 
- PO-.YY.-.-{supplier_name}-.-.#####
but that gives me
_Special Characters except "-", "#", "." and "/" not allowed in naming series_

sbkolate commented:
@raveslave It is merged in master

To use custom field, you need to make following naming series QTN-.YYYY.-.customer.-.#### where customer is custom field.

Please refer following link
https://discuss.erpnext.com/t/you-can-use-custom-field-in-naming-series-along-with-yy-mm-dd-option/10865
